core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -0,0 +1,387 @@
+
+    private static final List<String> MEMORIZED_JSON_TYPES = ImmutableList.of("array", "object");
+
+    /**
+     * Generate an array of json schema specifications for each method inside the service.
+     *
+     * @param serviceSpecification the service specification to generate the json schema from.
+     *
+     * @return ArrayNode that contains service specifications
+     */
+    static ArrayNode generate(ServiceSpecification serviceSpecification) {
+        // TODO: Test for Thrift and annotated services
+        final JsonSchemaGenerator generator = new JsonSchemaGenerator(serviceSpecification);
+        return generator.generate();
+    }
+
+    private final Set<ServiceInfo> serviceInfos;
+    private final Map<String, StructInfo> typeSignatureToStructMapping;
+    private final Map<String, EnumInfo> typeNameToEnumMapping;
+
+    private JsonSchemaGenerator(ServiceSpecification serviceSpecification) {
+        serviceInfos = serviceSpecification.services();
+        typeSignatureToStructMapping = serviceSpecification.structs().stream().collect(
+                Collectors.toMap(StructInfo::name, Function.identity()));
+        typeNameToEnumMapping = serviceSpecification.enums().stream().collect(
+                Collectors.toMap(EnumInfo::name, Function.identity()));
+    }
+
+    private ArrayNode generate() {
+        final ArrayNode definitions = mapper.createArrayNode();
+
+        final Set<ObjectNode> methodDefinitions = serviceInfos.stream().flatMap(
+                serviceInfo -> serviceInfo.methods().stream()
+                                          .map(this::generate)).collect(
+                toImmutableSet());
+
+        return definitions.addAll(methodDefinitions);
+    }
+
+    /**
+     * Generate the JSON Schema for the given {@link MethodInfo}.
+     *
+     * @param methodInfo the method to generate the JSON Schema for.
+     *
+     * @return ObjectNode containing the JSON schema for the parameter type.
+     */
+    private ObjectNode generate(MethodInfo methodInfo) {
+        final ObjectNode root = mapper.createObjectNode();
+
+        root.put("$id", methodInfo.id())
+            .put("title", methodInfo.name())
+            .put("description", methodInfo.descriptionInfo().docString())
+            .put("additionalProperties", false)
+            // TODO: Assumes every method takes an object, which is only valid for RPC based services
+            //  and most of the REST services.
+            .put("type", "object");
+
+        final List<FieldInfo> methodFields;
+        final Map<TypeSignature, String> visited = new HashMap<>();
+        final String currentPath = "#";
+
+        if (methodInfo.useParameterAsRoot()) {
+            final TypeSignature signature = methodInfo.parameters().get(0)
+                                                      .typeSignature();
+            final StructInfo structInfo = typeSignatureToStructMapping.get(signature.signature());
```

Comment:
Would it be solid to use `signature.name()`. A `signature` may change depending on how to represent the type.

Reply:
Signature names can overlap (pretty unlikely but I think it can happen) and in that case, our generated schema will be incorrect. I have taken a look at `equals` and `hashCode` methods of the Container type signature for example. That looked like a more comprehensive check.

At worst case scenario, we will generate some of the schema definitions redundantly. But for the other scenario, we can `$ref` to an incorrect type.